compiler/optimizer/OSRDefAnalysis.cpp

```diff
@@ -699,75 +699,280 @@ TR_OSRLiveRangeAnalysis::TR_OSRLiveRangeAnalysis(TR::OptimizationManager *manage
          {
-         traceMsg(comp(), "%s OSR live range analysis is not affordable for this method because of the expected cost\n",
-            optimizer()->getMethodSymbol()->signature(comp()->trMemory()));
-         traceMsg(comp(), "Returning...\n");
+         if (symRef->getSymbol()->isParm())
+            {
+            if (methodSymbol->sharesStackSlot(symRef))
+               {
+               sharesParm = true;
+               _sharedSymRefs.set(symRef->getReferenceNumber());
+               }
+            }
+         else if (symRef->getCPIndex() < methodSymbol->getFirstJitTempIndex())
+            {
+            containsAuto = true;
+            if (methodSymbol->sharesStackSlot(symRef))
+               _sharedSymRefs.set(symRef->getReferenceNumber());
+            }
          }
+      }
 
-      return 0;
+   TR_Array<List<TR::SymbolReference>> *pendingPushSymRefs = comp()->getMethodSymbol()->getPendingPushSymRefs();
+   for (int32_t i = 0; pendingPushSymRefs && i < pendingPushSymRefs->size(); ++i)
+      {
+      List<TR::SymbolReference> symRefsAtThisSlot = (*pendingPushSymRefs)[i];
+      ListIterator<TR::SymbolReference> symRefsIt(&symRefsAtThisSlot);
+      for (TR::SymbolReference *symRef = symRefsIt.getCurrent(); symRef; symRef = symRefsIt.getNext())
+         {
+         containsPendingPush = true;
+         _pendingPushSymRefs.set(symRef->getReferenceNumber());
+         if (comp()->getMethodSymbol()->sharesStackSlot(symRef))
+            _sharedSymRefs.set(symRef->getReferenceNumber());
+         }
       }
-   else if (comp()->isPeekingMethod())
+
+   if (comp()->getOption(TR_DisableOSRLiveRangeAnalysis))
       {
       if (comp()->getOption(TR_TraceOSR))
          {
-            traceMsg(comp(), "%s OSR live range analysis is not required because we are peeking\n",
-               optimizer()->getMethodSymbol()->signature(comp()->trMemory()));
-         traceMsg(comp(), "Returning...\n");
+         if (!_sharedSymRefs.isEmpty())
+            traceMsg(comp(), "OSRLiveRangeAnalysis is disabled but it is required for correctness. OSR transitions may not work.\n");
+         else
+            traceMsg(comp(), "OSRLiveRangeAnalysis is disabled.\n");
          }
       return 0;
       }
-   else if (!comp()->supportsInduceOSR())
+
+   // If pending push liveness was tracked in IlGen, process it before
+   // the full analysis, to cheapen it or avoid it entirely
+   if (containsPendingPush && comp()->pendingPushLivenessDuringIlgen())
       {
-      if (comp()->getOption(TR_TraceOSR))
+      partialAnalysis();
+      containsPendingPush = false;
+      }
+
+   if (containsAuto || sharesParm || containsPendingPush)
+      return fullAnalysis(sharesParm, containsPendingPush);
+
+   return 0;
+   }
+
+/*
+ * The partial analysis is only concerned with solving which pending push temps
+ * are live at certain OSR points.
+ */
+int32_t TR_OSRLiveRangeAnalysis::partialAnalysis()
```

Comment:
This feels like it needs more tracing... it logs the start, but not much else...

Reply:
Tracing has been added to the treetop walk and the method entry. The liveness and shared slot functions already had tracing.